rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+  attribute value, assume the `Target` denotes a pattern-based BYOC integration to explore. The `PartitionRule`
+  will import all the BYOC patterns and predicates automatically.
+- As above, but if global pattern has no matching entry, assume the `Target` denotes a predicate-based BYOC integration
+  to explore (eg `"tensorrt"`). The `PartitonRule` will look for and evaluate predicates with the
+  `"target.<compiler>"` attribute on all Relay operators.
+- Otherwise, assume the `Target` denotes a TVM-native target. The `PartitionRule` mimics `FuseOps`, but now generalized
+  to explore multiple candidates so as to leave room for possible BYOC candidates.
+
+Note that to make this approach work we need to allow for multiple `Target`s with the same `DLDeviceKind`. For the VM
+simply switching the `target` argument from dictionary to list and removing some redundant Python preprocessing code was
+all that was required to support this.
+
+The user can use `on_device` annotations to constrain sub-graphs to particular devices. When Collage is considering
+candidate partitions, it should be sure to choose a candidate `Target` which 'refines' the `Target` for every
+sub-expression discovered by the `PlanDevicesPass`. Given targets T and U we say 'T refines U' if T has a
+'"compiler"' and/or '"partition_rule"' attributes, U has no such attributes, and T and U otherwise agree on all other
+fields.
+
+### Phase 2
+
+Most of the hard work for this phase is carried by the `AllCandidates` implementations of the `PartitionRule`s. The main
+driver simply needs to index all the found `CandidatePartitions` by their minimum 'inside' `PostDfsIndex`
+for rapid retrieval during the shortest path search.
+
+### Phase 3
+
+We find it most natural to use Dijkstra to find the optimal partitioning. A `SearchState` is:
+
+- An `IndexSet` of the dataflow nodes already 'covered' by candidates on the best path to this state. This is the
+  identifying key for the state.
+- The predecessor `SearchState` in the best path to this state.
+- The `Cost` of the best path to this state. This is the order for the Dijkstra priority queue.
+- The `CandidatePartition` for the transition from the best predecessor to this state.
+
+The starting state has no covered nodes. The final state has all nodes covered.
+
+When expanding a state we could choose any `CandidatePartition` collected from phase 2 provided it doesn't overlap with
+the state's covered set. However, a search path applying candidates C then D is equivalent to one applying D then C, so
+we only consider candidates which intersect the next yet-to-be-covered dataflow node. For each such candidate we use
+the `CostEstimator` (with it's assumed cache) to get the candidate's cost, build the successor state, and 'relax' the
+successor state in the usual way. (See Appendix E for more details on `CostEstimator`.)
+
+The `HostPartitionRule` is used to allow some dataflow nodes to be 'left behind' for execution by the host.
```

Comment:
Why host treated differently?

Reply:
As the search proceeds we may encounter Relay subexpressions (eg an if) which no BYOC backend is interested in and which do not fit into TVM's notion of what can go in a kernel. At first I handled all of this inside the search algorithm, but then realized a 'HostPartitionRule' can move that special casing out of the core algorithm and into a very trivial rule implementation.

Reply:
Added an extra sentence here.